sklearn/neighbors/base.py

```diff
@@ -379,39 +433,45 @@ def kneighbors_graph(self, X, n_neighbors=None,
         --------
         NearestNeighbors.radius_neighbors_graph
         """
-        X = check_array(X, accept_sparse=['csr', 'csc', 'coo'])
-
         if n_neighbors is None:
             n_neighbors = self.n_neighbors
 
-        n_samples1 = X.shape[0]
+        # kneighbors does the None handling.
+        if X is not None:
+            X = check_array(X, accept_sparse='csr')
+            n_samples1 = X.shape[0]
+        else:
+            n_samples1 = self._fit_X.shape[0]
+
         n_samples2 = self._fit_X.shape[0]
         n_nonzero = n_samples1 * n_neighbors
         A_indptr = np.arange(0, n_nonzero + 1, n_neighbors)
 
         # construct CSR matrix representation of the k-NN graph
         if mode == 'connectivity':
-            A_data = np.ones((n_samples1, n_neighbors))
+            A_data = np.ones(n_samples1 * n_neighbors)
             A_ind = self.kneighbors(X, n_neighbors, return_distance=False)
 
         elif mode == 'distance':
-            data, ind = self.kneighbors(X, n_neighbors + 1,
-                                        return_distance=True)
-            A_data, A_ind = data[:, 1:], ind[:, 1:]
+            A_data, A_ind = self.kneighbors(X, n_neighbors, return_distance=True)
+            A_data = np.ravel(A_data)
 
         else:
             raise ValueError(
                 'Unsupported mode, must be one of "connectivity" '
                 'or "distance" but got "%s" instead' % mode)
 
-        return csr_matrix((A_data.ravel(), A_ind.ravel(), A_indptr),
-                          shape=(n_samples1, n_samples2))
+        kneighbors_graph = csr_matrix((
```

Comment:
This doesn't look like correct indentation. The first entry in the first argument tuple shouldn't be lining up with the kwarg name. What's wrong with:
```
        kneighbors_graph = csr_matrix((A_data, A_ind.ravel(), A_indptr),
                                      shape=(n_samples1, n_samples2))
```